stdlib/src/utils/static_tuple.mojo

```diff
@@ -254,3 +314,139 @@ struct StaticTuple[element_type: AnyRegType, size: Int](Sized):
+        self.array = __mlir_op.`kgen.undef`[_type = Self.type]()
+
+    @always_inline
+    fn __init__(inout self, *elems: Self.ElementType):
+        """Constructs a static tuple given a set of arguments.
+
+        Args:
+            elems: The element types.
+        """
+        _static_tuple_construction_checks[size]()
+        self = Self()
+        var i = 0
+        for e in elems:
+            self[i] = e[]
+            i += 1
+
+    @always_inline
+    fn __init__(inout self, values: VariadicList[Self.ElementType]):
+        """Creates a tuple constant using the specified values.
+
+        Args:
+            values: The list of values.
+        """
+        _static_tuple_construction_checks[size]()
+        # size is checked via the method call below
+        self.array = _create_array_collection_elem[size, Self.ElementType](
+            values
+        )
+
+    @always_inline("nodebug")
+    fn __len__(self) -> Int:
+        """Returns the length of the array. This is a known constant value.
+
+        Returns:
+            The size of the list.
+        """
+        return size
+
+    @always_inline
+    fn __refitem__[
+        mutability: __mlir_type.i1, self_life: AnyLifetime[mutability].type
+    ](
+        self: Reference[Self, mutability, self_life].mlir_ref_type, index: Int
+    ) -> Reference[Self.ElementType, mutability, self_life]:
+        """Gets a reference to the array element at the given index.
+
+        Parameters:
+            mutability: The inferred mutability of the type.
+            self_life: The inferred lifetime of the variant type.
+
+        Args:
+            index: The index of the element.
+
+        Returns:
+            A mutable reference to the element at the given index.
+        """
+        debug_assert(-size <= index < size, "index must be within bounds")
+        var normalized_idx = index
+
+        if index < 0:
+            normalized_idx += Reference(self)[].size
+
+        var array_copy = Reference(Reference(self)[].array)
+        var ptr = __mlir_op.`pop.array.gep`(
+            array_copy.get_unsafe_pointer().address, normalized_idx.value
+        )
+        return AnyPointer(ptr).__refitem__()
+
+    @always_inline
+    fn __refitem__[
+        mutability: __mlir_type.i1,
+        self_life: AnyLifetime[mutability].type,
+        index: Int,
+    ](self: Reference[Self, mutability, self_life].mlir_ref_type) -> Reference[
+        Self.ElementType, mutability, self_life
+    ]:
+        """Gets a reference to the array element at the given index.
+
+        Parameters:
+            mutability: The inferred mutability of the type.
+            self_life: The inferred lifetime of the variant type.
+            index: The index of the element.
+
+        Returns:
+            A mutable reference to the element at the given index.
+        """
+        constrained[-size <= index < size, "index must be within bounds"]()
+        var normalized_idx = index
+
+        if index < 0:
```

Comment:
**Suggestion** This could be run at compile time via `@parameter if` since `index` is a parameter (and not a runtime argument) in this context.  So, `normalized` index could be an `alias` too (we may need to define a smaller helper function and assign the result of calling to the `alias normalized_idx = compile_time_helper_function_to_compute_normalized_index()` for example).

After all that is said and done, what do you think about factoring out the "core logic" of the body of this function with the other `__refitem__` overload?  Worth it or nah?

Reply:
I factored out part of this to `_get_unsafe_reference` since the `gep` is really the part worth isolating IMO, and the main difference is that one can be a compile time `if`.